### DISCUSSION TITLE
Render USD asset with textures?

### DISCUSSION THREAD

**Discussion by cdwasden** [Autodesk/maya-usd#2272] [Q&A] [answered]
Hi all!

Our team is looking into USD for our future projects, we are seeing some things that feel very promising and exciting for us as a studio from your work!

One of our current holdups is trying to render assets with textures. Our assets currently use Arnold shaders. When I export a particular asset to USD using usdPreviewSurface materials, I can see in the textures showing up in the viewport - but when I render that USD file, the render has the simple solid colors that seem to be commonly associated with usdPreviewSurface:
![image](https://user-images.githubusercontent.com/3377907/162518903-c5769fcc-ddc0-40d6-849f-7b7e33750f5b.png)

We can use the Arnold-USD plugin to export these assets, and they will then render properly with textures... But we have found that these USD files don't seem to always play well with the USD Layer Editor, so they're not a perfect solution right now.

Are we missing something? Is there a way to export an USD asset from Maya such that you can bring that USD file in again and have it render with textures?

Thanks for your help!

**Comment by cdwasden** [comment, marked as answer]
We are still looking into this.

We've been updating to the latest version of the plugin and started testing with 2023. I have a co-worker that is able to generate USD files that correctly includes these materials and textures. Unfortunately, I am still unable to render these assets with textures.

Here is a test cube I am exporting, along with the settings I have when I export them:
![image](https://user-images.githubusercontent.com/3377907/170593551-ac44fca1-51c4-47df-be12-246350b6b09e.png)

I've got the Plug-In Configuration set to Arnold (sounds like that is new as of maya-usd 0.18), setting the materials to USD Preview Surface.

When I import this asset as a USD file into a fresh scene, I can see the textures in the viewport, but I do not see anything when I render inside of Arnold:
![image](https://user-images.githubusercontent.com/3377907/170593792-0614640b-e9d0-49fe-9c95-356d28cfdd23.png)

Am I missing something? If my co-worker can export and see something with textures but I can't, it sounds like something must be different between our machines... But I am at a loss.

**Reply by Lucas3Dspain** [reply]
I think(but I'm not sure) that the problem is the geomSubsets under pCube1.
Try the same thing with a torus or a sphere, IDK why but Maya creates subsets for the polyCube.

We are able to author assets from Maya, Houdini, and custom code, that renders properly with the textures (in hydra deegates). Both in Maya 2022.3 and 2023. mayaUsd 0.18 

If interested, I could share an example.

**Reply by cdwasden** [reply]
Thanks for the response, Lucas!

Unfortunately exporting with a sphere leads to the same results:
![image](https://user-images.githubusercontent.com/3377907/171249261-e01bf989-c972-4616-a7f2-c4b7baed48fa.png)


I've tried all kinds of combinations of export settings - turning things on, turning things off, importing it through the Layer Editor, importing through Bifrost. The only success I have had with rendering with textures is by using the ArnoldUSD option. That displays the textures when rendering in Arnold, although there are no colors displayed in the viewport anymore:
![image](https://user-images.githubusercontent.com/3377907/171259141-2c3c17ad-4d0c-4577-985c-e2ab51eba993.png)

I am not sure where to proceed with this. It does appear possible to render with textures using maya-usd, but I spent multiple weeks trying to figure this out with zero success.

Lucas, you mentioned sharing an example; I'd love to see what you have in case there is something there that helps!

**Reply by Lucas3Dspain** [reply]
Hi, 
I think your assets can render with textures if you render with the Arnold render delegate(hydra). 
You can do this from the viewport controls > renderer> Hydra Arnold . by default it is set to Viewport 2.0
![image](https://user-images.githubusercontent.com/25728227/171294485-64766c8d-dc24-421a-966e-d7a71bbb7115.png)


I'm sending you an example built from Houdini.
Custom code could be done to extract the same from Maya.

Make sure the zip is unzipped at C:\ so that the path of the asset is C:\usd_assets\patoLucas\patoLucas.usda
This is because Arnold doesn't like relative paths, or I couldn't make them work. :P

If you need to move the asset or change the Arnold image path you can change it from the patoLucas_mtl.usda file

Result of the example asset I send:
[usd_assets.zip](https://github.com/Autodesk/maya-usd/files/8809843/usd_assets.zip)
![maya_lDDAqO3iax](https://user-images.githubusercontent.com/25728227/171294564-f80e72d5-9111-4547-8194-0f85eb65a6c0.png)

**Reply by cdwasden** [reply]
Thank you so much for your help and the example files, Lucas! I've gleaned a lot from looking through them. I was able to get your test file to display correctly in the viewport and in the Arnold render view, just like your image above.

I did a side by side comparison between your USD files and the one that Maya generated, and I found some differences.

The Maya-USD file that I exported only generates the following material items (with relevant attributes):
A UsdUVTexture
A UsdPrimvarReader_float2
![image](https://user-images.githubusercontent.com/3377907/171518700-f2b2a54b-132f-4b3c-8258-362c384be6e8.png)

And then connects those to the Material. (Also, despite me setting the option to export a USDPreviewSurface material, there's no reference of that in the actual USD file.)

Your USD file has the following:
A UsdPreviewSurface
A UsdUVTexture (plugged into the USDPreviewSurface)
A UsdPrimvarReader_float2
A Shader that is defined as "arnold:standard_surface"
A shader that is defined as "arnold:image"

![image](https://user-images.githubusercontent.com/3377907/171518808-6930152a-9d66-4ee2-bc69-2b19b375cc83.png)

And then the material plugs in to the UsdPreviewSurface and the arnold:standard_surface shaders.

I edited my USD file to include those other shader nodes in the file. Once I did, I was able to see the object in the viewport and in the Arnold renderer:
![image](https://user-images.githubusercontent.com/3377907/171518199-fc9af50e-5cca-4f7a-ae9a-526c801429ea.png)

(I still need to figure out the messed up normals in the viewport, but at least I can test that on my own.)

A couple of notes:

1. When I edited the USD file (using a text editor) and then read the USD file in Maya, it seemed to keep a cache of the "old" USD file even though I had created a new Maya scene. I had to click this little "refresh" button on the Stage node to refresh the USD file:
![image](https://user-images.githubusercontent.com/3377907/171518398-47e01a19-4c1d-468e-bd82-90d841d88703.png)

2. Even though I couldn't get Maya to generate the Arnold shader nodes when exporting (as of 0.18), I was able to correctly view the texture in Arnold after manipulating the file, so the plugin is aware of the Arnold material setup at least. I'll be looking more into how to generate this on our end, but we can at least move forward with this.

Thanks again for all of your help, Lucas!!

**Comment by Lucas3Dspain** [comment]
Hi Chris, I'm glad this was helpful to you. But I wanted to clarify some things.

For the normals, you can try to set the subdivision scheme to none instead of catmullClark or add support edges.

The MTL layer structure is custom-made in Houdini to have it working as we needed. Can be easily replicated with the USD API.

The structure is made out of two shaders, one for the viewport  (usdPreviewSurface) and the other for Arnold (standardSurface).
The Arnold one is only available when using that render delegate. If we wanted to add Vray or RenderMan we would need to add another shader(and its node network) to have that rendering properly, while displaying the usdPreviewSurface in the viewport.
This is the beauty of USD, one asset that when properly set up can be rendered with multiple render engines. MaterialX tries to make this even simpler by only having one shader network that all the hydra render delegates understand (or at least that's my understanding).

Then we merge all the shaders into a material that switches between shaders depending on the renderer. This material is assigned to the prims.

The shaders/materials and the structure are not generated/exported from Maya, I used a Houdini setup to author these files.
![houdinifx_6GUw6iwooW](https://user-images.githubusercontent.com/25728227/171522934-4a252379-b54a-41c1-a231-6f24316abdd6.png)
![houdinifx_5Y86kXZXr6](https://user-images.githubusercontent.com/25728227/171522939-6ff41d47-922e-4c59-a854-5690089392f8.png)

In Maya, this shaders/material setup will look like this. But we didn't find a way to export that properly natively.
![maya_6CHCakqd8E](https://user-images.githubusercontent.com/25728227/171523252-f25d306d-58cd-4afd-94e2-914ce8d0aac2.png)

I don't know what kind of workflow are you seeking but if you just need to connect some textures to the shaders you could build a simple python script that builds a structure like the one I shared.
But if you want the shd artist to do complex shaders/lookdev, then the only way I see with publicly available tools is with Houdini's Solaris.

I hope that I explained myself haha
I'm happy to help if anything arises :)

**Reply by cdwasden** [reply]
That is all extremely helpful information - thank you, Lucas!!

**Comment by Ziethriel** [comment]
I just ran into this issue in Maya 2024, updating to the latest version of Arnold fixed it without having to adjust my content. Version 5.3.4.1
